### PR TITLE
Rewrote runtime condenser

### PR DESCRIPTION
**Pull request by DamianX**
### Differences from previous version
- It's written in Rust
- It actually sort of works

### Known bugs
- ~Doesn't know about infinite recursion errors~
- ~Doesn't always print the stack trace for infinite loops~

### Example
input.txt : http://game.ss13.moe/logs/Dev/runtime/2017-11-07-runtime.log
output.txt : https://gist.github.com/DamianX/be6cfeae1b407f0048f3687f6ad8a14d

### CODE REVIEW

File: tools/runtime-condenser/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+	runtimes
+}
+
+fn total_runtimes(runtimes: &Runtimes) -> usize {
+	runtimes.iter()
+		.fold(0, |count, (_ident, data)| count + data.counter)
+}
+
+fn total_unique_runtimes(runtimes: &Runtimes) -> usize {
+	runtimes.iter()
```
Comment (PJB3005) [inline tools/runtime-condenser/src/main.rs]: Can't you just do `runtimes.len()`?

File: tools/runtime-condenser/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+			}
+			LineKind::Junk => ()
+		}
+	}
+	runtimes
+}
+
+fn total_runtimes(runtimes: &Runtimes) -> usize {
+	runtimes.iter()
+		.fold(0, |count, (_ident, data)| count + data.counter)
```
Comment (PJB3005) [inline tools/runtime-condenser/src/main.rs]: Can't you use `sum()`?

File: tools/runtime-condenser/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+						continue;
+					}
+					data.details.push_str(&line);
+					data.details.push('\n');
+				}
+			}
+			LineKind::Runtime => {
+				let runtime_identifier = line[22..].to_owned();
+				current_state = State::Runtime(runtime_identifier.clone());
+				insert_runtime(&mut runtimes, runtime_identifier, RuntimeKind::RuntimeError);
```
Comment (PJB3005) [inline tools/runtime-condenser/src/main.rs]: You're cloning the identifier *twice* before checking whether that was even necessary (inside `insert_runtime`). It's all wasted if it only needed to increase the counter.

**Review by PJB3005 — DISMISSED** [review]

**Comment by PJB3005** [comment]
Something like JSON output could be nice.

**Comment by PJB3005** [comment]
Oh a good one: allow passing in and output files as command line arguments.

**Comment by DamianX** [comment]
from sood:

maximum stack limit:
```dm
runtime error: Maximum recursion level reached (perhaps there is an infinite loop)
To avoid this safety check, set world.loop_checks=0.
proc name: d (/proc/d)
  usr: Pseudonym (/mob)
  src: null
  usr.loc: null
  call stack:
d()
c()
b()
a()
o()
n()
m()
l()
k()
j()
...
a()
n()
l()
j()
h()
f()
d()
b()
Pseudonym (/mob): Login()
```
so everything after `call stack:` is a max of 20 lines
test for if the line is part of a new runtime or infinite recursion warning before trying to regex it for the proc name/args

**Comment by PJB3005** [comment]
https://github.com/DamianX/vgstation13/pull/2

**Comment by DamianX** [comment]
`rayon-hash` requires unstable features, but we don't really NEED it, we can use parallelization with the std hashmap just as well. It's just less efficient because it needs to `collect()` once. Do we move to nightly or do I get rid of `rayon-hash`?

**Comment by PJB3005** [comment]
Nightly I say.

**Comment by DamianX** [comment]
Apparently travis doesn't let you stick to a fixed nightly version (https://github.com/travis-ci/travis-ci/issues/7167)
This might one day cause the build to fail sporadically. Are we okay with this?

**Comment by PJB3005** [comment]
Eh in that case just remove it from Travis I'd say.

**Review by PJB3005 — APPROVED** [review]

**Pull request merged**